lib/codec.js

```diff
@@ -0,0 +1,153 @@
+}
+
+/**
+ * Gets a term codec map for the given set of JSON-LD Context URLs. This map
+ * can be used to perform encoding or decoding.
+ *
+ * @param {object} options - The arguments to the function.
+ * @param {Array} options.contextUrls - The JSON-LD Context URLs to use when
+ *   building the term codec map.
+ * @param {object} options.documentLoader - The document loader to use when
+ *   dereferencing URLs.
+ * @param {object} [options.appTermMap] - A map of JSON-LD terms and
+ *   their associated CBOR-LD term codecs.
+ * @param {boolean} [options.encode] - If true, a map of term encoding codecs
+ *   are returned.
+ * @param {boolean} [options.decode] - If true, a map of term decoding codecs
+ *   are returned.
+ *
+ * @returns {object} - The term codec map for performing encoding or decoding.
+ */
+export async function getTermCodecMap({
+  contextUrls, appTermMap = new Map(), documentLoader, encode, decode
+} = {}) {
+  const termCodecMap = new Map();
+
+  // retrieve all JSON-LD Contexts
+  const contexts = await Promise.all(contextUrls.map(
+    async contextUrl => JSON.parse(await documentLoader(contextUrl))));
+
+  // generate a term codec map for each context
+  contexts.forEach(context => {
+    const map = _generateTermEncodingMap(context);
+    map.forEach((value, key) => termCodecMap.set(key, value));
+  });
+
+  // calculate the final values for the term codec map
+  // FIXME: ensure this is the same in all locales; may need to
+  // pick a particular locale to ensure it/do something else
+  const terms = Array.from(termCodecMap.keys()).sort();
+  terms.forEach((term, index) => {
+    if(encode) {
+      termCodecMap.get(term).value = index;
+    } else if(decode) {
+      termCodecMap.get(term).value = term;
+      termCodecMap.set(index, termCodecMap.get(term));
+    }
+  });
+
+  // overlay the app term map on the existing term encoding map
+  if(decode && appTermMap.size > 0) {
+    const codecMap = new CodecMapCodec();
+    codecMap.set({value: appTermMap, termCodecMap});
+    appTermMap = codecMap.decodeCBOR();
+  }
+
+  // clear terms used during encoding
+  if(decode) {
+    terms.forEach(term => termCodecMap.delete(term));
+  }
+
+  for(const [key, value] of appTermMap.entries()) {
+    const codec = (encode) ? termCodecs.get(value) : value;
+    if(codec === undefined) {
+      throw new CborldError(
+        'ERR_UNKNOWN_TERM_CODEC',
+        `Unknown codec '${value}' specified for term '${key}'.`);
+    }
+    termCodecMap.get(key).codec = codec;
+  }
+
+  return termCodecMap;
+}
+
+// Generates a term encoding map given a JSON-LD Context
+function _generateTermEncodingMap(context) {
+  const termCodecMap = new Map();
+
+  // sanity check that context is a valid JSON-LD context object
+  if(!context['@context']) {
+    throw new CborldError(
+      'ERR_INVALID_JSONLD_CONTEXT',
+      'The JSON-LD Context does not contain a top-level @context value: ' +
+      JSON.stringify(context));
```

Comment:
Hrm, at this point in the code, we've long lost the URL and keeping it around to this point in the code would be a pain. I wanted them to see the entire context if they hit this error since it may be deep in the bowels of a software system. I have updated the error to only print out the first 100 characters or so, hopefully that helps.

Fixed in 8defe20338aedabfb7964f1a801b6a8fbcaf40bb.